File: functions/CopBET_degree_distribution_entropy.m
% out = CopBET_degree_distribution_entropy(in,varargin)
%
% Copenhagen Brain Entropy Toolbox: Degree distribution entropy
% Evaluates temporal entropy as in Viol et al., 2017. For each session, the
% correlation coefficient matrix is established, where coefficients with a
% non-significant p-value are set to zero. Then, the matrix is binarised
% according a large range of absolute value thresholds and the mean degree
% is calculated for each threshold. Then, for a range of desired mean
% degrees, the degree distribution of the corresponding binarised
% matrix is calculated, and finally, the Shannon entropy is returned. 
%
% Input:
%   in: a matrix (nxp,n>1) or a table where the first column contains
%   matrices (in cells), e.g., different subjects or scan sessions.
%   
%   varargin (name-value pairs):
%   keepdata: Indicates whether the output table also should contain the
%   input data, i.e., by adding an extra column containing entropy values.
%   Defaults to true
%   parallel: Whether to run temporal entropy in parallel (true)
%
%
% Neurobiology Research Unit, 2023
% Please cite McCulloch, Olsen et al., 2023: "Navigating Chaos in
% Psychedelic Neuroimaging: A Rigorous Empirical Evaluation of the Entropic
% Brain Hypothesis" if you use CopBET in your studies. Please read the
% paper to get a notion of our recommendations regarding the use of the
% specific methodologies in the toolbox.

% ASO March-April 2023

% Copyright (C) 2023 Anders Stevnhoved Olsen & Drummond E-Wen McCulloch
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.
%
%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see http://www.gnu.org/licenses/.

function out = CopBET_degree_distribution_entropy(in,varargin)

[out,numworkers,in] = CopBET_function_init(in,varargin);

% Select ranges and initialize variables
opts.thresh=0.01:0.001:0.99; %threshold values
opts.chosen_degrees = 1:100; %degrees across which to test
opts.bins = 0.5:100.5; %histogram bins

%load data
parfor(ses = 1:height(in),numworkers)
    disp(['Working on entropy calculations for session: ',num2str(ses)])
    ts = in{ses,1}{1};
    [R,PVAL] = corr(ts);
    sensible_data_check(R,['correlation matrix, session ',num2str(ses)]);
    R(PVAL>0.05) = 0;
    entropy{ses} = calc_viol17_entropy(R,opts);
    sensible_data_check(entropy{ses},['entropy, session ',num2str(ses)]);
    
end
out.entropy = entropy';
end

%% functions

% test graph
% graph = [0,1,1,0,1,1,1,0,0;1,0,1,1,0,0,0,0,0;1,1,0,0,1,0,0,0,0;0,1,0,0,0,0,0,0,0;1,0,1,0,0,0,0,0,0;1,0,0,0,0,0,1,0,0;1,0,0,0,0,1,0,1,1;0,0,0,0,0,0,1,0,1;0,0,0,0,0,0,1,1,0]

function entropy_out = calc_viol17_entropy(Z,opts)
bin_matrix=cell(1,length(opts.thresh));
meandegree = zeros(length(bin_matrix),1);
entropy_out = zeros(length(opts.chosen_degrees),1);
for tt=1:length(opts.thresh) %Loop through threshold values
    
    %binarise the matrix
    bin_matrix{tt}=abs(Z)>opts.thresh(tt);
    
    %Calculate mean degree
    meandegree(tt) = mean(degrees_und(bin_matrix{tt}));
    
end

%Extract the matrix with each appropriate degree
for i=1:length(opts.chosen_degrees)
    [~,idx] = min(abs(meandegree - opts.chosen_degrees(i)));
    [Prob,~] = histcounts(degrees_und(bin_matrix{idx}),opts.bins,'Normalization','probability');
    %calculate the Shannon entropy of degree distribution
    entropy_out(i) = nansum(-Prob.*log(Prob));
end
end

File: copbet_py/functions/CopBET_degree_distribution_entropy.py
"""
CopBET_degree_distribution_entropy
=====================================
Copenhagen Brain Entropy Toolbox: Degree distribution entropy.

Evaluates temporal entropy as in Viol et al., 2017.
For each session, the Pearson correlation matrix is computed and entries
with non-significant p-values (p > 0.05) are set to zero. The matrix is
then binarized across a range of thresholds to yield a sweep of mean
degrees. For a set of desired mean degrees, the degree distribution is
computed and its Shannon entropy returned.

Input
-----
sessions : list of np.ndarray, each shape (T, N)

Returns
-------
entropy : list of np.ndarray, each shape (n_degrees,)
    Degree-distribution entropy across target degrees, per session.

Reference
---------
Viol et al., 2017.

Please cite McCulloch, Olsen et al., 2023 if you use CopBET.
"""

import numpy as np
from scipy import stats
import bct


def CopBET_degree_distribution_entropy(sessions,
                                       thresholds=None,
                                       chosen_degrees=None):
    """
    Compute degree-distribution Shannon entropy.

    Parameters
    ----------
    sessions : list of np.ndarray, each shape (T, N)
    thresholds : array-like or None
        Absolute value thresholds to sweep (default: 0.01 to 0.99, step 0.001).
    chosen_degrees : array-like or None
        Target mean degrees to evaluate (default: 1 to 100).

    Returns
    -------
    entropy : list of np.ndarray, each shape (n_chosen_degrees,)
    """
    if thresholds is None:
        thresholds = np.arange(0.01, 0.991, 0.001)
    if chosen_degrees is None:
        chosen_degrees = np.arange(1, 101)

    entropy = []

    for ses, ts in enumerate(sessions):
        print(f"Processing session {ses + 1}/{len(sessions)}...", end='\r')
        ts = np.asarray(ts, dtype=float)
        T, N = ts.shape

        # Pearson correlation + p-values
        R, pval = _corrcoef_pval(ts)

        # Zero out non-significant correlations
        R[pval > 0.05] = 0.0

        ent = _calc_viol17_entropy(R, thresholds, chosen_degrees)
        entropy.append(ent)

    return entropy


def _corrcoef_pval(ts):
    """Compute Pearson correlation matrix and matrix of p-values."""
    T, N = ts.shape
    R = np.corrcoef(ts.T)
    pval = np.ones((N, N))

    for i in range(N):
        for j in range(i + 1, N):
            r = R[i, j]
            # Two-tailed t-test: t = r * sqrt(T-2) / sqrt(1-r^2)
            r_clamped = np.clip(r, -1 + 1e-10, 1 - 1e-10)
            t_stat = r_clamped * np.sqrt(T - 2) / np.sqrt(1 - r_clamped ** 2)
            p = 2 * stats.t.sf(np.abs(t_stat), df=T - 2)
            pval[i, j] = p
            pval[j, i] = p

    np.fill_diagonal(pval, 0.0)
    return R, pval


def _calc_viol17_entropy(R, thresholds, chosen_degrees):
    """
    Compute degree-distribution Shannon entropy for a range of mean degrees.

    Parameters
    ----------
    R : np.ndarray, shape (N, N)
        (Thresholded) correlation matrix.
    thresholds : array-like
        Absolute value thresholds to sweep.
    chosen_degrees : array-like
        Target mean degrees.

    Returns
    -------
    entropy_out : np.ndarray, shape (len(chosen_degrees),)
    """
    N = R.shape[0]
    bins = np.arange(0.5, N + 1.5)  # degree bins 1..N

    # Binarize at each threshold and record mean degree
    bin_matrices = []
    mean_degrees = np.zeros(len(thresholds))
    for tt, thr in enumerate(thresholds):
        B = (np.abs(R) > thr).astype(float)
        np.fill_diagonal(B, 0)
        deg = bct.degrees_und(B)
        mean_degrees[tt] = np.mean(deg)
        bin_matrices.append(B)

    # For each desired mean degree, find nearest threshold and compute entropy
    entropy_out = np.zeros(len(chosen_degrees))
    for i, target_deg in enumerate(chosen_degrees):
        idx = np.argmin(np.abs(mean_degrees - target_deg))
        B = bin_matrices[idx]
        deg = bct.degrees_und(B)
        prob, _ = np.histogram(deg, bins=bins, density=False)
        prob = prob / prob.sum() if prob.sum() > 0 else prob
        mask = prob > 0
        entropy_out[i] = -np.sum(prob[mask] * np.log(prob[mask]))

    return entropy_out
